### PR TITLE
forge: introduce webpack plugin and remove parcel

### PR DESCRIPTION
**Pull request by aryanshridhar**
Introduces forge+webpack configuration and removes parcel for the bundling process.

<strong>TODOs</strong>
* Handle the management of assets files in the webpack configuration.
* Figure out the webpack management for `process.env.NODE_ENV` in main process.
* Fix the current issue/current blocker (listed below)

<strong>Current blocker - </strong>

```
Uncaught ReferenceError: require is not defined
    events external node-commonjs "events":1
    Webpack 12
external node-commonjs "events":1
```

Corresponding unanswered (according to our use case) thread: https://stackoverflow.com/questions/66112475/electron-forge-with-webpack-getting-require-is-not-defined-when-importing-eve.

### CODE REVIEW

File: forge.config.js [hunk 1]
```diff
@@ -1,6 +1,7 @@
 const path = require('path');
 const fs = require('fs');
 const packageJson = require('./package.json');
+const root = process.cwd();
```
Comment (dsanders11) [inline forge.config.js:4]: Very minor nit, but can we move this down to be in the section on lines 6-7 instead? Usually the section of requires at the top of the file only contains requires.

Comment (aryanshridhar) [reply forge.config.js:4]: Cool!

File: tools/webpack/webpack.renderer.config.js [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+});
+
+// Fonts
+rules.push({
+  test: /\.(eot|ttf|woff2?)(\?v=\d+\.\d+\.\d+)?$/i,
+  type: 'asset/resource',
+  generator: {
+    filename: 'fonts/[name][ext]',
+  },
+});
```
Comment (dsanders11) [inline tools/webpack/webpack.renderer.config.js:66]: Let's change these. The current code, pushing into the array exported by the other modules, it actually mutating the array value. That's generally not a good idea as it can introduce subtle bugs (any other import after this one will see the mutated value).

Let's use spread syntax to accomplish this without mutating `rules` and `plugins`. It would look something like this:

```
module.exports = {
  module: {
    rules: [
        ..rules,
        {
            test: /\.(eot|ttf|woff2?)(\?v=\d+\.\d+\.\d+)?$/i,
            type: 'asset/resource',
            generator: {
                filename: 'fonts/[name][ext]',
            },
            // etc
        }
    ]
  },
  plugins: [
      ...plugins,
      new MonacoWebpackPlugin({
          languages: ['typescript', 'javascript', 'html', 'css'],
      }),
      // etc
  ]
```

**Review by dsanders11 — CHANGES_REQUESTED** [review]

File: src/main/windows.ts [hunk 3]
```diff
@@ -50,7 +57,9 @@ export function createMainWindow(): Electron.BrowserWindow {
   console.log(`Creating main window`);
   let browserWindow: BrowserWindow | null;
   browserWindow = new BrowserWindow(getMainWindowOptions());
-  browserWindow.loadFile('./dist/static/index.html');
+  browserWindow.loadURL(
+    !!process.env.JEST ? './dist/static/index.html' : MAIN_WINDOW_WEBPACK_ENTRY,
```
Comment (dsanders11) [inline src/main/windows.ts:61]: I think this should actually be `./.webpack/renderer/main_window/index.html`... I don't think the tests are actually making usage of it, but let's change it to that.

Same with `preload` above, let's change that to `./.webpack/renderer/main_window/preload.js`.

File: tools/webpack/webpack.renderer.config.js (hunk 2 above)
Comment (aryanshridhar) [reply tools/webpack/webpack.renderer.config.js:66]: Yup, Sounds right! I was only doing that in only renderer process as it's the last piece of code that executes :)

File: src/renderer/templates.ts [hunk 4]
```diff
@@ -2,17 +2,17 @@ import * as path from 'path';
...
  * @returns {Promise<EditorValues>}
  */
 export function getTemplateValues(name: string): Promise<EditorValues> {
-  const folder = path.join(
-    __dirname,
-    '../../static/show-me',
-    name.toLowerCase(),
-  );
-  return readFiddle(folder);
+  // TODO: Figure out a better method to handle template path
```
Comment (dsanders11) [inline src/renderer/templates.ts:14]: Let's drop this comment.

File: src/renderer/components/settings-credits.tsx [hunk 5]
```diff
@@ -19,9 +18,9 @@ interface Contributor {
   api: string;
   login: string;
   avatar: string;
-  name: string;
-  bio: string;
-  location: string;
+  name: string | null;
+  bio: string | null;
+  location: string | null;
```
Comment (dsanders11) [inline src/renderer/components/settings-credits.tsx:23]: Are these required changes to make things work, or just updating this to be more accurate? If the latter, let's not include it in this PR, can be done in a follow-up. Since this is already a large PR, I'd like to keep things focused.

File: forge.config.js [hunk 6]
```diff
@@ -1,9 +1,11 @@
 const path = require('path');
 const fs = require('fs');
 const packageJson = require('./package.json');
+const contributors = require('./tools/contributors');
```
Comment (dsanders11) [inline forge.config.js:4]: nit: `const { maybeFetchContributors } = require('./tools/contributors');` and avoid the awkward `contributors` name.

File: src/renderer/components/settings-credits.tsx (hunk 5 above)
Comment (aryanshridhar) [reply src/renderer/components/settings-credits.tsx:23]: I had to add these as within test `mockContributorsBroken` is defined holding these values as `null` as well (It gives an type error when calling the method - `getContributors`)

Comment (aryanshridhar) [reply src/renderer/components/settings-credits.tsx:23]: But now with using `wrapper.setState({ contributors })`, This isn't required, so I will be reverting this :)

File: .github/workflows/build.yaml [hunk 7]
```diff
@@ -28,6 +28,12 @@ jobs:
           cache: yarn
       - name: Install
         run: yarn --frozen-lockfile --network-timeout 100000 || yarn --frozen-lockfile --network-timeout 100000 || yarn --frozen-lockfile --network-timeout 100000
+      - name: Create fake contributors
+        uses: 1arp/create-a-file-action@0.2
+        with:
+          path: 'static'
+          file: 'contributors.json'
+          content: "[]"
```
Comment (dsanders11) [inline .github/workflows/build.yaml:36]: Just adding a note here to investigate why `tests/globalSetup.js` doesn't work for CI for creating this file. 👍

**Review by VerteDinde — APPROVED** [review]
Nice job!
